### PR TITLE
Add additional information when we throw an 'unreachable' exception

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
We've been getting "unreachable" exceptions that we're finding to difficult to take action on due to trouble determining what the actual throw point of hte exception was.  This adds that information to the exception, in a similar way to how we added information to our Contract calls so that devs investigating the issue could track the exception back to the exact location it was thrown.

### CODE REVIEW

File: src/Compilers/Core/Portable/InternalUtilities/ExceptionUtilities.cs [hunk 1]
```diff
@@ -24,10 +25,8 @@ internal static Exception UnexpectedValue(object? o)
             return new InvalidOperationException(output);
         }
 
-        internal static Exception Unreachable
-        {
-            get { return new InvalidOperationException("This program location is thought to be unreachable."); }
-        }
+        internal static Exception Unreachable([CallerFilePath] string? path = null, [CallerLineNumber] int line = 0)
+            => new InvalidOperationException($"This program location is thought to be unreachable. File='{path}' Line={line}");
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/InternalUtilities/ExceptionUtilities.cs:29]: this is the actual update to the helper property to be a helper method.

**Review by jasonmalinowski — APPROVED** [review]
Signing off on the helper change; and obviously not looking at the remaining 455 files. :smile:

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-compiler for this change to help diagnosability of watson dumps for exceptions.

**Review by genlu — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
ack.  did not intend to merge this.

**Comment by CyrusNajmabadi** [comment]
Talking to infra now.

**Review by jcouv — COMMENTED** [review]
LGTM Thanks (iteration 1)

**Review by 333fred — COMMENTED** [review]
Compiler changes LGTM

**Pull request merged**